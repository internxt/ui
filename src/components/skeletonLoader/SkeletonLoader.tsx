interface SkeletonLoaderItemProps {
  skeletonItem: Array<React.ReactElement> | undefined;
  columns: Array<string>;
}

interface SkeletonLoaderProps {
  skeleton: Array<SkeletonLoaderItemProps>;
}

export const SkeletonLoaderItem = ({ skeletonItem, columns }: SkeletonLoaderItemProps): React.ReactElement => {
  return (
    <div
      data-testid="skeleton-loader-item"
      className={'group relative flex h-14 animate-pulse flex-row items-center pl-14 pr-5'}
    >
      {new Array(columns.length).fill(0).map((col, i) => (
        <div
          key={`${col}-${i}`}
          className={`relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 ${columns[i]}`}
        >
          {skeletonItem?.[i]}
        </div>
      ))}
    </div>
  );
};

/**
 * SkeletonLoader component to display loading placeholders in list layouts.
 *
 * @param skeleton - Array of skeleton items, {skeletonItem, columns}
 *                 - The params of each sekeletonItem:
 *                            - skeletonItem: - Array of elements representing the loading placeholders.
 *                            - columns: - An array of CSS class names for each column, defining their layout and style.
 *
 */
const SkeletonLoader = ({ skeleton }: SkeletonLoaderProps): React.ReactElement => {
  return (
    <>
      {new Array(skeleton.length).fill(0).map((_col, i) => (
        <SkeletonLoaderItem
          key={`skinSkeletonRow${i}`}
          skeletonItem={skeleton[i].skeletonItem}
          columns={skeleton[i].columns}
        />
      ))}
    </>
  );
};

export default SkeletonLoader;
