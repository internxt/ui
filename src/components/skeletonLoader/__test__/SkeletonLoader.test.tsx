import React from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import SkeletonLoader from '../../../components/skeletonLoader/SkeletonLoader';
import { expect, describe, it } from 'vitest';

describe('SkeletonLoader', () => {
  const renderSkeletonLoader = () => {
    const skeletonData = new Array(3).fill(0).map((_row, index) => ({
      skeletonItem: [
        <div key={`skin-skeleton-1-${index}`} className="mr-3 flex w-full flex-row items-center space-x-4">
          <div className="h-8 w-8 rounded-md bg-gray-5" />
          <div className="h-4 w-full rounded bg-gray-5" />
        </div>,
        <div key={`skin-skeleton-2-${index}`} className="h-4 w-64 rounded bg-gray-5" />,
        <div key={`skin-skeleton-3-${index}`} className="ml-3 h-4 w-24 rounded bg-gray-5" />,
      ],
      columns: ['flex grow items-center min-w-driveNameHeader', 'w-date', 'w-size'],
    }));
    return render(<SkeletonLoader skeleton={skeletonData} />);
  };

  it('should match snapshot', () => {
    const skeletonLoader = renderSkeletonLoader();
    expect(skeletonLoader).toMatchSnapshot();
  });

  it('should render the correct number of SkeletonLoaderItem', () => {
    renderSkeletonLoader();

    const skeletonItems = screen.getAllByTestId('skeleton-loader-item');
    expect(skeletonItems).toHaveLength(3);
  });

  it('should render columns with the correct classes', () => {
    renderSkeletonLoader();

    const items = screen.getAllByTestId('skeleton-loader-item')[0].querySelectorAll('div');

    const column1 = items[0];
    const column2 = items[4];
    const column3 = items[6];
    expect(column1).toHaveClass('flex');
    expect(column2).toHaveClass('w-date');
    expect(column3).toHaveClass('w-size');
  });

  it('should handle an empty skeleton array correctly', () => {
    render(<SkeletonLoader skeleton={[]} />);

    const skeletonItems = screen.queryAllByTestId('skeleton-loader-item');
    expect(skeletonItems).toHaveLength(0);
  });
});
