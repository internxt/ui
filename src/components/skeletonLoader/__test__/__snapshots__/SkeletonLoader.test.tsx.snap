// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SkeletonLoader > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="group relative flex h-14 animate-pulse flex-row items-center pl-14 pr-5"
        data-testid="skeleton-loader-item"
      >
        <div
          class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 flex grow items-center min-w-driveNameHeader"
        >
          <div
            class="mr-3 flex w-full flex-row items-center space-x-4"
          >
            <div
              class="h-8 w-8 rounded-md bg-gray-5"
            />
            <div
              class="h-4 w-full rounded bg-gray-5"
            />
          </div>
        </div>
        <div
          class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 w-date"
        >
          <div
            class="h-4 w-64 rounded bg-gray-5"
          />
        </div>
        <div
          class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 w-size"
        >
          <div
            class="ml-3 h-4 w-24 rounded bg-gray-5"
          />
        </div>
      </div>
      <div
        class="group relative flex h-14 animate-pulse flex-row items-center pl-14 pr-5"
        data-testid="skeleton-loader-item"
      >
        <div
          class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 flex grow items-center min-w-driveNameHeader"
        >
          <div
            class="mr-3 flex w-full flex-row items-center space-x-4"
          >
            <div
              class="h-8 w-8 rounded-md bg-gray-5"
            />
            <div
              class="h-4 w-full rounded bg-gray-5"
            />
          </div>
        </div>
        <div
          class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 w-date"
        >
          <div
            class="h-4 w-64 rounded bg-gray-5"
          />
        </div>
        <div
          class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 w-size"
        >
          <div
            class="ml-3 h-4 w-24 rounded bg-gray-5"
          />
        </div>
      </div>
      <div
        class="group relative flex h-14 animate-pulse flex-row items-center pl-14 pr-5"
        data-testid="skeleton-loader-item"
      >
        <div
          class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 flex grow items-center min-w-driveNameHeader"
        >
          <div
            class="mr-3 flex w-full flex-row items-center space-x-4"
          >
            <div
              class="h-8 w-8 rounded-md bg-gray-5"
            />
            <div
              class="h-4 w-full rounded bg-gray-5"
            />
          </div>
        </div>
        <div
          class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 w-date"
        >
          <div
            class="h-4 w-64 rounded bg-gray-5"
          />
        </div>
        <div
          class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 w-size"
        >
          <div
            class="ml-3 h-4 w-24 rounded bg-gray-5"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="group relative flex h-14 animate-pulse flex-row items-center pl-14 pr-5"
      data-testid="skeleton-loader-item"
    >
      <div
        class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 flex grow items-center min-w-driveNameHeader"
      >
        <div
          class="mr-3 flex w-full flex-row items-center space-x-4"
        >
          <div
            class="h-8 w-8 rounded-md bg-gray-5"
          />
          <div
            class="h-4 w-full rounded bg-gray-5"
          />
        </div>
      </div>
      <div
        class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 w-date"
      >
        <div
          class="h-4 w-64 rounded bg-gray-5"
        />
      </div>
      <div
        class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 w-size"
      >
        <div
          class="ml-3 h-4 w-24 rounded bg-gray-5"
        />
      </div>
    </div>
    <div
      class="group relative flex h-14 animate-pulse flex-row items-center pl-14 pr-5"
      data-testid="skeleton-loader-item"
    >
      <div
        class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 flex grow items-center min-w-driveNameHeader"
      >
        <div
          class="mr-3 flex w-full flex-row items-center space-x-4"
        >
          <div
            class="h-8 w-8 rounded-md bg-gray-5"
          />
          <div
            class="h-4 w-full rounded bg-gray-5"
          />
        </div>
      </div>
      <div
        class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 w-date"
      >
        <div
          class="h-4 w-64 rounded bg-gray-5"
        />
      </div>
      <div
        class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 w-size"
      >
        <div
          class="ml-3 h-4 w-24 rounded bg-gray-5"
        />
      </div>
    </div>
    <div
      class="group relative flex h-14 animate-pulse flex-row items-center pl-14 pr-5"
      data-testid="skeleton-loader-item"
    >
      <div
        class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 flex grow items-center min-w-driveNameHeader"
      >
        <div
          class="mr-3 flex w-full flex-row items-center space-x-4"
        >
          <div
            class="h-8 w-8 rounded-md bg-gray-5"
          />
          <div
            class="h-4 w-full rounded bg-gray-5"
          />
        </div>
      </div>
      <div
        class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 w-date"
      >
        <div
          class="h-4 w-64 rounded bg-gray-5"
        />
      </div>
      <div
        class="relative flex h-full shrink-0 flex-row items-center overflow-hidden whitespace-nowrap border-b border-gray-5 w-size"
      >
        <div
          class="ml-3 h-4 w-24 rounded bg-gray-5"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
